pkg/controller/kubedirectorcluster/util.go

```diff
@@ -35,3 +41,147 @@ func updateSchedulingErrorMessage(
 		}
 	}
 }
+
+// RunConfigScript executes an app configuration script
+// with different notifications which describe the current cluster state
+func RunConfigScript(
+	reqLogger logr.Logger,
+	cr *kdv1.KubeDirectorCluster,
+	roleName string,
+	podName string,
+	configArg ConfigArg,
+	expectedContainerID string,
+	loggingErr bool,
+) error {
+
+	cmd := GetAppConfigCmd(expectedContainerID, configArg)
+	cmdErr := executor.RunScript(
+		reqLogger,
+		cr,
+		cr.Namespace,
+		podName,
+		expectedContainerID,
+		executor.AppContainerName,
+		"app config",
+		strings.NewReader(cmd),
+	)
+	if loggingErr && cmdErr != nil {
+		shared.LogErrorf(
+			reqLogger,
+			cmdErr,
+			cr,
+			shared.EventReasonMember,
+			"failed to run startscript with --{%s} in member{%s} in role{%s}",
+			string(configArg),
+			podName,
+			roleName,
+		)
+	}
+	return cmdErr
+}
+
+// QueueNotify prepares the info for handling a lifecycle event to a currently
+// ready node, and adds the info to the node's notification queue. We are
+// notifying about new members either being added to the modifiedRole (if it
+// has members in creating state) or being removed (if it has members in
+// delete pending state).
+func QueueNotify(
+	reqLogger logr.Logger,
+	cr *kdv1.KubeDirectorCluster,
+	podName string,
+	roleName string,
+	// This function should return the operation argument as first value and updated FQDNs as second value
+	evalOpFqdnsFn func() (string, string),
+) {
+
+	op, deltaFqdns := evalOpFqdnsFn()
+	if deltaFqdns == "" && (op == "addnodes" || op == "delnodes") {
+		// No nodes actually being created/deleted. One example of this
+		// is in the creating case where none have been successfully
+		// configured.
+		return
+	}
+	// Notify the node iff the event is registered during initial configuration.
+	appCr, appErr := catalog.GetApp(cr)
+	if appErr != nil {
+		shared.LogError(
+			reqLogger,
+			appErr,
+			cr,
+			shared.EventReasonCluster,
+			"app referenced by cluster does not exist")
+	}
+	role := catalog.GetRoleFromID(appCr, roleName)
+
+	if role.EventList != nil && !shared.StringInList(op, *role.EventList) {
+		return
+	}
+
+	shared.LogInfof(
+		reqLogger,
+		cr,
+		shared.EventReasonNoEvent,
+		"will notify member{%s}: %s",
+		podName,
+		op,
+	)
+	// Compose the notify command arguments.
+	arguments := []string{
+		"--" + op,
+		"--nodegroup 1", // currently only 1 nodegroup possible
+		"--role",
+		roleName,
+		"--fqdns",
+		deltaFqdns,
+	}
+	notifyDesc := kdv1.NotificationDesc{
+		Arguments: arguments,
+	}
+
+	for i, role := range cr.Status.Roles {
```

Comment:
This looping in QueueNotify will work, but it makes me sad. :-)

I think we can avoid this by passing in a pointer to the member. That means that in the caller, instead of using `range` to iterate over the members, it will have to iterate over the count of members (let's say that is `j`) and get each member by its array subscript (e.g. `Members[j]`) ... then the caller will be working with the actual member object (not a copy) and can pass its pointer to QueueNotify.

There's various examples of that approach in the code. `range` is a real troublemaker.

Reply:
> This looping in QueueNotify will work, but it makes me sad. :-)
> 
> I think we can avoid this by passing in a pointer to the member. That means that in the caller, instead of using `range` to iterate over the members, it will have to iterate over the count of members (let's say that is `j`) and get each member by its array subscript (e.g. `Members[j]`) ... then the caller will be working with the actual member object (not a copy) and can pass its pointer to QueueNotify.
> 
> There's various examples of that approach in the code. `range` is a real troublemaker.

Well, if I understood correctly, I replaced it to `for j := 0; j < len(role.Members); j++ {...}` loop form